net/graphsync_fetcher.go

```diff
@@ -185,3 +224,45 @@ func (gsf *GraphSyncFetcher) loadTipsetFromCids(ctx context.Context, cids []cid.
 
 	return types.NewTipSet(validatedBlocks...)
 }
+
+func (gsf *GraphSyncFetcher) missingCids(cids []cid.Cid) ([]cid.Cid, error) {
+	missing := make([]cid.Cid, 0, len(cids))
+	for _, cid := range cids {
+		hasBlock, err := gsf.store.Has(cid)
+		if err != nil {
+			return nil, err
+		}
+		if !hasBlock {
+			missing = append(missing, cid)
+		}
+	}
+	return missing, nil
+}
+
+type requestPeerFinder struct {
+	peerTracker graphsyncFallbackPeerTracker
+	currentPeer peer.ID
+	triedPeers  map[peer.ID]struct{}
+}
+
+func newRequestPeerFinder(peerTracker graphsyncFallbackPeerTracker, from peer.ID) *requestPeerFinder {
+	pri := &requestPeerFinder{peerTracker, from, make(map[peer.ID]struct{})}
+	pri.triedPeers[from] = struct{}{}
+	return pri
+}
+
+func (pri *requestPeerFinder) CurrentPeer() peer.ID {
+	return pri.currentPeer
+}
+
+func (pri *requestPeerFinder) FindNextPeer() error {
```

Comment:
how does the peer tracker get its list?  any reason not to cache/memoize this  list in requestPeerFinder just for the current invocation?

Reply:
according to @ZenGround0 it could change on each call.